File: core/src/commonMain/kotlin/work/socialhub/kslack/api/methods/request/files/remote/FilesRemoteAddRequest.java
package com.github.seratch.jslack.api.methods.request.files.remote;

import com.github.seratch.jslack.api.methods.SlackApiRequest;

/**
 * https://api.slack.com/methods/files.remote.add
 */
public class FilesRemoteAddRequest implements SlackApiRequest {

    /**
     * Authentication token. Requires scope: `remote_files:write`
     */
    private String token;

    /**
     * Creator defined GUID for the file.
     */
    private String externalId;

    /**
     * URL of the remote file.
     */
    private String externalUrl;

    /**
     * Title of the file being shared.
     */
    private String title;

    /**
     * type of file
     */
    private String filetype;

    /**
     * File containing contents that can be used to improve searchability for the remote file.
     */
    private byte[] indexableFileContents;

    /**
     * Preview of the document via multipart/form-data.
     */
    private byte[] previewImage;

    FilesRemoteAddRequest(String token, String externalId, String externalUrl, String title, String filetype, byte[] indexableFileContents, byte[] previewImage) {
        this.token = token;
        this.externalId = externalId;
        this.externalUrl = externalUrl;
        this.title = title;
        this.filetype = filetype;
        this.indexableFileContents = indexableFileContents;
        this.previewImage = previewImage;
    }

    public static FilesRemoteAddRequestBuilder builder() {
        return new FilesRemoteAddRequestBuilder();
    }

    public String getToken() {
        return this.token;
    }

    public String getExternalId() {
        return this.externalId;
    }

    public String getExternalUrl() {
        return this.externalUrl;
    }

    public String getTitle() {
        return this.title;
    }

    public String getFiletype() {
        return this.filetype;
    }

    public byte[] getIndexableFileContents() {
        return this.indexableFileContents;
    }

    public byte[] getPreviewImage() {
        return this.previewImage;
    }

    public void setToken(String token) {
        this.token = token;
    }

    public void setExternalId(String externalId) {
        this.externalId = externalId;
    }

    public void setExternalUrl(String externalUrl) {
        this.externalUrl = externalUrl;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setFiletype(String filetype) {
        this.filetype = filetype;
    }

    public void setIndexableFileContents(byte[] indexableFileContents) {
        this.indexableFileContents = indexableFileContents;
    }

    public void setPreviewImage(byte[] previewImage) {
        this.previewImage = previewImage;
    }

    public static class FilesRemoteAddRequestBuilder {
        private String token;
        private String externalId;
        private String externalUrl;
        private String title;
        private String filetype;
        private byte[] indexableFileContents;
        private byte[] previewImage;

        FilesRemoteAddRequestBuilder() {
        }

        public FilesRemoteAddRequest.FilesRemoteAddRequestBuilder token(String token) {
            this.token = token;
            return this;
        }

        public FilesRemoteAddRequest.FilesRemoteAddRequestBuilder externalId(String externalId) {
            this.externalId = externalId;
            return this;
        }

        public FilesRemoteAddRequest.FilesRemoteAddRequestBuilder externalUrl(String externalUrl) {
            this.externalUrl = externalUrl;
            return this;
        }

        public FilesRemoteAddRequest.FilesRemoteAddRequestBuilder title(String title) {
            this.title = title;
            return this;
        }

        public FilesRemoteAddRequest.FilesRemoteAddRequestBuilder filetype(String filetype) {
            this.filetype = filetype;
            return this;
        }

        public FilesRemoteAddRequest.FilesRemoteAddRequestBuilder indexableFileContents(byte[] indexableFileContents) {
            this.indexableFileContents = indexableFileContents;
            return this;
        }

        public FilesRemoteAddRequest.FilesRemoteAddRequestBuilder previewImage(byte[] previewImage) {
            this.previewImage = previewImage;
            return this;
        }

        public FilesRemoteAddRequest build() {
            return new FilesRemoteAddRequest(token, externalId, externalUrl, title, filetype, indexableFileContents, previewImage);
        }

        public String toString() {
            return "FilesRemoteAddRequest.FilesRemoteAddRequestBuilder(token=" + this.token + ", externalId=" + this.externalId + ", externalUrl=" + this.externalUrl + ", title=" + this.title + ", filetype=" + this.filetype + ", indexableFileContents=" + java.util.Arrays.toString(this.indexableFileContents) + ", previewImage=" + java.util.Arrays.toString(this.previewImage) + ")";
        }
    }
}

File: core/src/commonMain/kotlin/work/socialhub/kslack/api/methods/request/files/remote/FilesRemoteAddRequest.kt
package work.socialhub.kslack.api.methods.request.files.remote

import work.socialhub.kslack.api.methods.FormRequest
import work.socialhub.kslack.api.methods.SlackApiRequest
import kotlin.js.JsExport

/**
 * https://api.slack.com/methods/files.remote.add
 */
@JsExport
class FilesRemoteAddRequest(
    /** Authentication token. Requires scope: `remote_files:write` */
    override var token: String?,
    /** Creator defined GUID for the file. */
    var externalId: String?,
    /** URL of the remote file. */
    var externalUrl: String?,
    /** Title of the file being shared. */
    var title: String?,
    /** Type of file */
    var filetype: String?,
    /** File containing contents that can be used to improve searchability for the remote file. */
    var indexableFileContents: ByteArray,
    /** Preview of the document via multipart/form-data. */
    var previewImage: ByteArray
) : SlackApiRequest, FormRequest {

    override fun toMap(): Map<String, Any> {
        return mutableMapOf<String, Any>().also {

            it.addParam("external_id", externalId)
            it.addParam("external_url", externalUrl)
            it.addParam("title", title)
            it.addParam("filetype", filetype)

            /*
            TODO
            if (IndexableFileContents() != null) {
                RequestBody indexableFileContents = RequestBody.create(Filetype() != null ? MediaType.parse(Filetype) : null, IndexableFileContents);
                form.addFormDataPart("indexable_file_contents", Title(), indexableFileContents);
            }
            if (PreviewImage() != null) {
                RequestBody previewImage = RequestBody.create(Filetype() != null ? MediaType.parse(Filetype) : null, PreviewImage);
                form.addFormDataPart("preview_image", Title(), previewImage);
            }
            */
        }
    }
}